docs: required documentation (#73)
PascalHonegger/ChaosConnect: README.md

The modern, distributed and scalable implementation of a game inspired by Connect Four.

# Gameplay

## Rules

> - There are two main fractions: Yellow vs Red

Reviewer: ```suggestion
- There are two main factions: Yellow vs Red
```

Author: Done

The frontend is written in [Svelte](https://svelte.dev/).
As it's main purpose is to display the current state of the board, we decided that frameworks such as Angular are overkill.

### Load balancing
We use [Envoy](https://www.envoyproxy.io/) as a reverse proxy to handle load balancing.
We had to use Envoy as it's the only reverse proxy which currently supports grpc-web.

### Backend
The backend is split into two parts:
* Scaling: Communicates directly with our Frontend, issues and validates JWT, caches game state and sends game updates to all clients
* Central: Manages the actual game state, synchronizes requests and handles game logic, stores persistent information, such as user credentials and scores, in a json document

Both backends use [Micronaut](https://micronaut.io/) with [Kotlin](https://kotlinlang.org/).

### Storage
We decided not to use a database but instead store the (very minimalistic) data in a JSON document.

## Communication
Bidirectional communication is enabled through [gRPC](https://grpc.io/).
For example, this allows `Rohan` to send a game update event to all `Joestar` instances, which then forward them realtime to all `Doppio` clients.
TODO:
- gRPC:

>   - Authentication: Not used because of lack of support by Svelte ecosystem (overlap with [Symmetric JWT](#symmetric-jwt)?)?

Author: According to [Wikipedia](url),

> gRPC supports the usage of TLS and token based authentication

.
This was not an option for us, but I don't remember why.

Reviewer: We couldn't easily use the micronaut-security package because the feature is still WIP for gRPC (see https://github.com/micronaut-projects/micronaut-grpc/issues/164).
The official token based authentication works by using Google as a token provider, but we didn't want to have a vendor lock-in. 
In the end we decided to use simple symmetric tokens because of the project scope. The client does use the metadata to send the token which was recommended back in 2018 by the grpc-web team (https://github.com/grpc/grpc-web/issues/207#issuecomment-406134504)

Author: Done

>   - Bidirectional streaming and control flow: Updates only? How bidirectional is it?

Author: Is anything else streamed besides game state updates / events?

Reviewer: No, the requests are synchronous. The updates are streamed from Rohan to Joestar and then from Joestar to Doppio

Author: Done. I mentioned the kinds of control flow (streams, request-based) and the almost synchronous programming style offered by Kotlin and TypeScript.

>   - Blocking or non-blocking bindings: Relevant for programming (`suspend fun`s, `Promise`s + `async` + `await`)?

Author: I think that asynchronicity was handled nicely thanks to `suspend fun`s and `Promise`s + `async` + `await`.
What do you think?

Reviewer: Yes the whole request from start to finish is non-blocking thanks to Kotlin coroutines & JavaScript Promises

Author: Done

>   - Cancellation and timeouts: Relevant for programming (cancellation and timeout exceptions)?

Author: Cancellations and timeouts are mapped to exceptions, I guess.
`Flow` seems to offer convenient retry and completion hooks, as seen in `ch.chaosconnect.joestar.services.RohanService#getGameUpdates`.

Author: Done, kinda. I just mentioned callback methods for stream representations in general.

  - More?
- JWT termination on Joestar servers
- More?

## Configuration

> TODO:
> - Locality
> - Processing
> - More?

Reviewer: ```suggestion
We use [Micronaut Application Configuration](https://docs.micronaut.io/latest/guide/index.html#config) to make some parts of the application configurable. These configurations can be set through the `application.yml` file at compile time or throught an environment variable at runtime. The following configurations are probably the most interesting ones to configure, for a full list see the source code:

- Joestar
  - `JOESTAR_PORT`: The port at which Rohan is listening
  - `ROHAN_SERVER`: The hostname or ip of the Rohan server
  - `ROHAN_PORT`: The port at which Rohan is listening
  - `JWT_SECRET`: The base-64 encoded 512-bit private key used for signing the JWT
- Rohan
  - `ROHAN_PORT`: The port at which Rohan is listening

```

Author: Done

## Maintainability
TODO:
- Testing: 200+ unit tests

> - More?

Reviewer: We also use state of the art technology and are using the latest (!!) version of everything (e.g. Kotlin released a week ago, targeting JDK16)

Author: Done

# Infrastructure

## Docker setup

> TODO: Docker for everything

Reviewer: ```suggestion
In order to enable easy deployment every service is dockerized. We do not use docker containers for developing, but you can easily build the containers locally to test their cross-container communication. We also use docker to run gRPC code generation for grpc-web to ensure the code gets generated with the same compiler version on every device.

The docker images are generally optimized for file size and try to use the smallest available base image.

Our images also implement [Docker Healthchecks](https://docs.docker.com/engine/reference/builder/#healthcheck) which can be used to determine if a server is irrecoverably broken.
```

Author: Done

## Reverse Proxy
TODO:
- Responsibilities (overlap with [Separation of concerns](#separation-of-concerns)?)
- Load Balancing strategy
- More?

## CI/CD
TODO:
- GitHub actions

> - More?

Author: Done

> ## PKI
> TODO:
> - Certbot
> - More?
> 
> ## Hosting
> TODO:
> - Linode
> - More?

Reviewer: ```suggestion
## Hosting

At the time of writing this, the latest stable version of ChaosConnect is deployed at [chaos.honegger.dev](https://chaos.honegger.dev/). We decided to use the cloud provider [Linode](https://www.linode.com/) because they provide fair pricing and a good free initial credit.

Setting up hosting was very easy, you just copy the docker-compose.yml file, replace the placeholder jwt-secret with a generated one and you're almost ready to start. If you don't already have a valid certificate around, you can easily generate one using [Certbot](https://certbot.eff.org/) using a command similar to the following and mounting them to their corresponding location within the Speedwagon container:

`sudo docker run -it --rm --name certbot -v "/etc/letsencrypt:/etc/letsencrypt" -v "/var/lib/letsencrypt:/var/lib/letsencrypt" -p 80:80 certbot/certbot certonly --standalone`
```

Author: Done

# Design Decisions

## gRPC
TODO: Rationale
- Fundamental requirement as part of the assignment

> - More?

Reviewer: I wanted real time updates and gRPC provides a way to stream real time updates. Implementing real time update from server-to-server and server-to-browser is no easy task and implementing a type safe variant throught websockes would've taken a lot longer than gRPC.

Author: Done

## Micronaut
TODO: Rationale
- Dependency injection

> - More?

Reviewer: - Supports gRPC out of the box
- Configuration Management
- Promises fast startup and performance throught build-time dependency injection

Author: Done

## Kotlin
TODO: Rationale
- Conciseness
- `null` safety

> - More?

Author: Done

## Svelte
TODO: Rationale
- Light-weight framework (suitable for small projects)
- More?

## Envoy
TODO: Rationale

> - Compatibility (gRPC, JWT)?

Reviewer: Envoy is only reverse proxy with official and native grpc-web support

Author: Done

> - More?

Reviewer: Envoy does round robin load balancing and only considers servers for load balancing which fulfill a readiness check. For example, a joestar instance without a valid rohan connection is not considered ready and will not be load-balanced

Author: Done

## Symmetric JWT
TODO: Rationale
- Compatibility (gRPC)?
- Library support (Svelte ecosystem)?

> - More?

Author: Done

## Notifications
TODO: Rationale

> - Why real time updates with change events instead of whole state?

Author: Done

>   - More people => make playing field slightly higher

Author: Done

  - Fewer people => Mark (now superfluous) rows and columns as soon to be deleted, delete after a certain time passed
- If a disk you placed is part of a 4 disk long line you get points, and the associated columns are cleared

## Future Plans

> - You get more points if you're on the losing fraction for balance purposes

Reviewer: ```suggestion
- You get more points if you're on the losing faction for balance purposes
```

Author: Done

As it's main purpose is to display the current state of the board, we decided that frameworks such as Angular are overkill.

### Load balancing

We use [Envoy](https://www.envoyproxy.io/) as a reverse proxy to handle load balancing.
We had to use Envoy as it's the only reverse proxy which currently supports `grpc-web`.

### Backend
The backend is split into two parts:
* Scaling: Communicates directly with our Frontend, issues and validates JWT, caches game state and sends game updates to all clients
* Central: Manages the actual game state, synchronizes requests and handles game logic, stores persistent information, such as user credentials and scores, in a json document

Both backends use [Micronaut](https://micronaut.io/) with [Kotlin](https://kotlinlang.org/).

### Storage
We decided not to use a database but instead store the (very minimalistic) data in a JSON document.

## Communication

### Bi-directionality

Bidirectional communication is enabled through [gRPC](https://grpc.io/).
For example, this allows `Rohan` to send a game update event to all `Joestar` instances, which then forward them realtime to all `Doppio` clients.

### Authentication

> We use a custom solution based on symmetric JWT for authentication, as we did not want to commit to the standard solution provided by the Micronaut framework (see [Symmetric JWT](#symmetric-jwt)).

Reviewer: ```suggestion
Because we use gRPC for communication, which is based on HTTP, we send the authentication token in the [Authorization Header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization). We use a custom solution based on symmetric JWT for authentication, as we did not want to commit to a vendor-specific or work-in-progress solution (see [Symmetric JWT](#symmetric-jwt)).
```

Author: Done

| Streams                | `kotlinx.coroutines.flow.Flow` + callback methods           | `ClientReadableStream` + callback methods |
| 'Synchronous' Requests | `suspend fun` + `kotlinx.coroutines.BuildersKt.runBlocking` | `Promise` + `async` + `await`             |

### JWT

JWTs are issued and terminated by Joestar servers.

## Configuration

We use [Micronaut Application Configuration](https://docs.micronaut.io/latest/guide/index.html#config) to make some parts of the application configurable.
These configurations can be set through the `application.yml` file at compile time or through an environment variable at runtime.
The following configurations are probably the most interesting ones to configure, for a full list see the source code:

- Joestar
  - `JOESTAR_PORT`: The port at which Rohan is listening
  - `ROHAN_SERVER`: The hostname or ip of the Rohan server
  - `ROHAN_PORT`: The port at which Rohan is listening
  - `JWT_SECRET`: The base-64 encoded 512-bit private key used for signing the JWT
- Rohan
  - `ROHAN_PORT`: The port at which Rohan is listening

## Maintainability

> We use state-of-the-art technology and are using the latest (!) version of everything (e. g. Kotlin version released a week ago, targeting JDK 16).

Reviewer: ```suggestion
We use state-of-the-art technology and are using the latest version to take advantage of the latest language and framework features. For example, many Kotlin Coroutine features we rely upon to easily implement real time updates were released as stable May 14th, just in time for the v1 release.
```

Author: Done

We try to avoid unnecessary code cohesion and favor testability instead.
Currently, we have more than 200 unit tests.

# Infrastructure

## Docker setup

In order to enable easy deployment every service is dockerized.
We do not use docker containers for developing, but you can easily build the containers locally to test their cross-container communication.
We also use docker to run gRPC code generation for `grpc-web` to ensure the code gets generated with the same compiler version on every device.

The docker images are generally optimized for file size and try to use the smallest available base image.

Our images also implement [Docker Healthchecks](https://docs.docker.com/engine/reference/builder/#healthcheck) which can be used to determine if a server is irrecoverably broken.

## Reverse Proxy

> We use a dockerized Envoy reverse proxy server as a load balancer.

Reviewer: ```suggestion
We use a dockerized Envoy reverse proxy server as a load balancer (see [Envoy](#envoy) for details).
```

Author: Done

For example, a Joestar instance without a valid rohan connection is not considered ready and will not be load-balanced.

## Symmetric JWT

We couldn't easily use the `micronaut-security` package because the feature is still WIP for gRPC (see [`micronaut-grpc` issue #164](https://github.com/micronaut-projects/micronaut-grpc/issues/164)).
The official token-based authentication works by using Google as a token provider, but we didn't want to have a vendor lock-in.

In the end, we decided to use simple symmetric tokens because of the project scope.

The client does use the metadata to send the token [which was recommended back in 2018 by the `grpc-web` team](https://github.com/grpc/grpc-web/issues/207#issuecomment-406134504).

## Notifications

In order to preserve network bandwidth, which can be pricey depending on the hosting provider, we use real-time updates with light-weight change events instead of complete states.

# Operation
The easiest way to get ChaosConnect running is using docker-compose. We do not provider support for running the software components otherwise.

> ## Prerequisites

Reviewer: Maybe also document that for local development you need to have JDK16, Node 16 and a modern version of docker installed. Our project also contains run configurations for IntelliJ, which is our IDE of choice.

Author: Done
